pkg/provisioner/ironic/dependencies.go

```diff
@@ -0,0 +1,112 @@
+package ironic
+
+import (
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/drivers"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+)
+
+const (
+	checkRequeueDelay = time.Second * 10
+)
+
+type ironicDependenciesChecker struct {
+	client    *gophercloud.ServiceClient
+	inspector *gophercloud.ServiceClient
+}
+
+func newIronicDependenciesChecker(client *gophercloud.ServiceClient, inspector *gophercloud.ServiceClient) *ironicDependenciesChecker {
+	return &ironicDependenciesChecker{
+		client:    client,
+		inspector: inspector,
+	}
+}
+
+func (i *ironicDependenciesChecker) IsReady() (result provisioner.Result, err error) {
+
+	ready, err := i.checkIronic()
+	if ready && err == nil {
+		ready, err = i.checkIronicInspector()
+	}
+
+	if !ready && err == nil {
+		result.Dirty = true
+		result.RequeueAfter = checkRequeueDelay
+	}
+
+	return result, err
+}
+
+func (i *ironicDependenciesChecker) hasUnexpectedResponseCode(err error) bool {
+
+	errType := reflect.TypeOf(err)
+
+	if errType.Kind() == reflect.Struct {
```

Comment:
I don't understand what this block is doing. What error types need to be handled as special cases like this?

Reply:
The `gophercloud.ServiceClient.Get` returns an error if the HTTP status code is not present in a predefined [list](https://github.com/gophercloud/gophercloud/blob/d48e557b9093ff51c3dae00405d891200934ba8c/provider_client.go#L306) (by default it contains just 200 for a get). Every distinct error code has a distinct type, and all of them are just embedding the `gophercloud.ErrUnexpectedResponseCode` struct type. This block tries to detect if the current `error` interface embeds such embedded type. An alternative will be to use a type switch and enumerate all the possible types (maybe more readable but also slightly more error prone). Another interesting alternative approach could be to use error unwrapping convention introduced since [Golang1.13](https://blog.golang.org/go1.13-errors) but it would require a signficant refactoring of the code.

Reply:
I reviewed the implementation via type switch, a little bit verbose but looks more readable, and seems to catch all the cases better.